### PR TITLE
Add ARM Cortex-M1 DesignStart FPGA support

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
This PR adds support for the ARM Cortex-M1 DesignStart FPGA reference design running on the Digilent Arty development board. The PR consists of the following additions:

- ARM Cortex-M1 CPU support. The Cortex-M1 is basically a Cortex-M0 optimised for implementation within an FPGA.
- Minimalistic ARM DesignStart SoC support. Although this is not a SoC per se, it provides an easy way for users to use Zephyr on a Cortex-M1 implemented within an FPGA with just a board definition specific to the given FPGA logic design. The SoC support is prepared for addition of other ARM DesignStart variants (e.g. Cortex-M3).
- Xilinx UART Lite serial driver with interrupt support
- Xilinx AXI GPIO driver
- A board definition for the Cortex-M1 ARM DesignStart FPGA reference design running on the Digilent Arty (Xilinx Artix-7 based) FPGA development board.

Enhancements which will come after this PR has been accepted:
- Support for the ARM V2C-DAPLINK board used in the ARM DesignStart FPGA reference design

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
Sanitycheck passes for this board:

```
$ ./scripts/sanitycheck -p arty_a7_arm_designstart_m1 --device-testing --device-serial /dev/ttyUSB1
Renaming output directory to /home/brix/Projects/zephyrproject/zephyr/sanity-out.3
INFO    - JOBS: 2
INFO    - Building initial testcase list...
INFO    - 868 test configurations selected, 0 configurations discarded due to filters.

Device testing on:

| Platform                   | ID   | Serial device   |
|----------------------------|------|-----------------|
| arty_a7_arm_designstart_m1 |      | /dev/ttyUSB1    |

INFO    - Adding tasks to the queue...
INFO    - Total complete:  187/ 187  100%  skipped:  718, failed:    0
INFO    - 150 of 150 tests passed (100.00%), 0 failed, 718 skipped with 0 warnings in 901.60 seconds
INFO    - In total 959 test cases were executed on 1 out of total 273 platforms (0.37%)
INFO    - 150 tests executed on platforms, 0 tests were only built.

Hardware distribution summary:

| Board                      | ID   |   Counter |
|----------------------------|------|-----------|
| arty_a7_arm_designstart_m1 |      |       150 |
```

**Comment by nashif** [comment]
any documentation for the board and how to use it with Cortex-M1?

**Comment by henrikbrixandersen** [comment]
> any documentation for the board and how to use it with Cortex-M1?

@nashif Yes. I will write the documentation tonight and move this PR out of draft.

**Comment by nashif** [comment]
> @nashif Yes. I will write the documentation tonight and move this PR out of draft.

thanks, I have this board but never used it for anything, so maybe I will try this guy :)

**Comment by henrikbrixandersen** [comment]
> thanks, I have this board but never used it for anything, so maybe I will try this guy :)

@nashif Documentation is now included. Please give it a spin!

**Comment by henrikbrixandersen** [comment]
> @nashif Documentation is now included. Please give it a spin!

@nashif I have thrown in a GPIO driver for toying with as well.

**Comment by henrikbrixandersen** [comment]
The Xilinx AXI GPIO driver included in this PR has passed the `gpio_basic_api` test. The test is not directly usable on the reference design, since no suitable GPIOs are available. However, on a slightly modified FPGA logic design (with and additional GPIO controller), the test passes:

```
*** Booting Zephyr OS build zephyr-v2.3.0-1719-g5298874b9e87  ***
Running test suite gpio_basic_test
===================================================================
START - test_gpio_port
Validate device GPIO_2
Check GPIO_2 output 0 connected to input 1
OUT 0 to IN 1 linkage works
- bits_physical
- pin_physical
- check_raw_output_levels
- check_logic_output_levels
- check_input_levels
- bits_logical
- check_pulls
NOTE: cannot configure pin as disconnected; trying as input
pull-up not supported
 PASS - test_gpio_port
===================================================================
START - test_gpio_callback_add_remove
test_callback_add_remove not supported
 PASS - test_gpio_callback_add_remove
===================================================================
START - test_gpio_callback_self_remove
test_callback_self_remove not supported
 PASS - test_gpio_callback_self_remove
===================================================================
START - test_gpio_callback_enable_disable
test_callback_enable_disable not supported
 PASS - test_gpio_callback_enable_disable
===================================================================
START - test_gpio_callback_variants
interrupts not supported
interrupts not supported
interrupts not supported
interrupts not supported
interrupts not supported
interrupts not supported
interrupts not supported
interrupts not supported
interrupts not supported
 PASS - test_gpio_callback_variants
===================================================================
Test suite gpio_basic_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL

```

File: boards/arm/arty/arty_a7_arm_designstart_m1.dts [hunk 1]
```diff
@@ -0,0 +1,225 @@
+/*
```
Comment (ioannisg) [inline boards/arm/arty/arty_a7_arm_designstart_m1.dts:1]: I'd rather see this definition under dts/arm/_vendor_ , at least the part that defines the SoC (e.g. not the LEDs). That's the convention at least, and it is followed in most Cortex-m SoCs.

File: tests/arch/arm/arm_interrupt/src/arm_interrupt.c [hunk 2]
```diff
@@ -95,9 +95,15 @@ void test_arm_interrupt(void)
 
 			if (NVIC_GetPendingIRQ(i)) {
 				/* If the NVIC line is pending, it is
-				 * guaranteed that it is implemented.
+				 * guaranteed that it is implemented; clear the
```
Comment (ioannisg) [inline tests/arch/arm/arm_interrupt/src/arm_interrupt.c:98]: Can you elaborate, why this modification is required? (in this and in similar test suites)

File: tests/arch/arm/arm_no_multithreading/src/main.c [hunk 3]
```diff
@@ -78,9 +78,16 @@ void test_main(void)
...
-				 * guaranteed that it is implemented.
+				 * guaranteed that it is implemented; clear the
+				 * line.
 				 */
-				break;
+				NVIC_ClearPendingIRQ(i);
+
+				if (!NVIC_GetPendingIRQ(i)) {
+					/* The NVIC line is cleared; use it. */
+					NVIC_SetPendingIRQ(i);
```
Comment (ioannisg) [inline tests/arch/arm/arm_no_multithreading/src/main.c:88]: here seems you are pending the line again, so why did you need to un-pend it earlier?

**Review by ioannisg — CHANGES_REQUESTED** [review]
thanks for the work @henrikbrixandersen 
(Would be great to have in the future, also QEMU coverage)
for now, i think i've spotted some minor issues and I've got a request of a proper .dtsi header for the SoC.

Please, explain also the NVIC-related modifications in the test suites.

File: boards/arm/arty/doc/index.rst [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+The Spartan-7 and Artix-7 based Arty board do not contain a CPU, but require a
+so-called soft processor to be instantiated within the FPGA in order to run
+Zephyr. The Zynq-7000 based Arty boards are not yet supported by Zephyr.
+
+ARM Cortex-M1 DesignStart FPGA
+******************************
+
+One such soft processor design is the Cortex-M1 `ARM DesignStart FPGA`_ Xilinx
+edition reference design from ARM. This design targets the Spartan-7 or Artix-7
+based Arty boards. Zephyr only support the Artix-7 based boards for now.
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply boards/arm/arty/doc/index.rst]: Fixed, thanks (and thanks for reviewing!).

File: arch/arm/core/aarch32/cortex_m/Kconfig [hunk 5]
```diff
@@ -23,6 +23,14 @@ config CPU_CORTEX_M0PLUS
 	help
 	  This option signifies the use of a Cortex-M0+ CPU
 
+config CPU_CORTEX_M1
+	bool
+	select CPU_CORTEX_M
+	select ARMV6_M_ARMV8_M_BASELINE
+	select CPU_CORTEX_M_HAS_SYSTICK
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply arch/arm/core/aarch32/cortex_m/Kconfig:30]: The SysTick timer is enabled when selecting `OS Extensions` under the Cortex-M1 customisation options. These include extensions include both the SysTick, SVC, and the NVIC. I am not sure it is worth attempting to support a Cortex-M1 with `OS extensions` disabled.

File: boards/arm/arty/arty_a7_arm_designstart_m1.dts (hunk 1 above)
Comment (henrikbrixandersen) [reply boards/arm/arty/arty_a7_arm_designstart_m1.dts:1]: `arm` is both an architecture and a vendor here, similar to the other SoCs provided by ARM Limited. The soft-core SoC is provided by ARM, but it can be implemented on silicon from various vendors. ARM Limited supports implementing it on at least Xilinx and GOWIN semiconductors FPGAs for now.

File: tests/arch/arm/arm_interrupt/src/arm_interrupt.c (hunk 2 above)
Comment (henrikbrixandersen) [reply tests/arch/arm/arm_interrupt/src/arm_interrupt.c:98]: A NVIC IRQ line can be used as a level-detect non-interrupt signal. IRQ 7 (top-most IRQ on this design) is implemented as such.

Previously the tests simply verified that an unused IRQ line could be set to pending, but not that it could also be cleared. This causes problems when the given IRQ line is used as a level-detect non-interrupt signal, since the input value is latched (and since it may already have been pending when we started the routine to locate an unused IRQ line, depending on the input level supplied to that line).

The modification extends the detection logic to ensure that an unused IRQ line can both be set pending and cleared again. This makes the tests pass for this design as the detection logic will now skip IRQ 7 (the level-detect signal) and select IRQ 6 and downwards.

File: tests/arch/arm/arm_no_multithreading/src/main.c (hunk 3 above)
Comment (henrikbrixandersen) [reply tests/arch/arm/arm_no_multithreading/src/main.c:88]: See above.

**Comment by henrikbrixandersen** [comment]
> thanks for the work @henrikbrixandersen

@ioannisg You are welcome. Thank you for reviewing.

> (Would be great to have in the future, also QEMU coverage)

Agreed.

> for now, i think i've spotted some minor issues and I've got a request of a proper .dtsi header for the SoC.

That DTSI header for the SoC would only contain this (as the rest are specific to what logic and peripherals are implemented in the given FPGA design):

```
/dts-v1/;
#include <arm/armv6-m.dtsi>

/ {

};

&nvic {
	arm,num-irq-priority-bits = <2>;
};

```
 
For the upcoming ARM Cortex-M3 DesignStart FPGA support, it won't even contain the `arm,num-irq-priority-bits` since that is specified by the FPGA design as well.

This is similar as to how the other SoCs from ARM Limited (mps2 and musca) are defined within Zephyr. They too let the board DTS define the entire devicetree except for the CPU part, which is included as a DTSI.

> Please, explain also the NVIC-related modifications in the test suites.

Please see above.

File: arch/arm/core/aarch32/cortex_m/Kconfig (hunk 5 above)
[the remark this answers is not in the document]
Comment (ioannisg) [reply arch/arm/core/aarch32/cortex_m/Kconfig:30]: It does not matter, IMO. We cannot list the HAS_SYSTICK as a capability of Cortex-M1, since it is optional. Adding it will create confusion on which features are always-present, and it will require re-write as soon as there's an FPGA without the systick.

File: tests/arch/arm/arm_interrupt/src/arm_interrupt.c (hunk 2 above)
Comment (ioannisg) [reply tests/arch/arm/arm_interrupt/src/arm_interrupt.c:98]: Thanks for the explanation. I think you need to add some of this explanation in the logic, as a comment, to make it easy for the next reader. Sthg like: "if the NVIC line can be successfully un-pended, it is guaranteed that it can be used for software interrupt triggering" or similar...

File: arch/arm/core/aarch32/swap_helper.S [hunk 6]
```diff
@@ -170,8 +170,8 @@ out_fp_endif:
     /* Restore previous interrupt disable state (irq_lock key)
      * (We clear the arch.basepri field after restoring state)
      */
-#if (defined(CONFIG_CPU_CORTEX_M0PLUS) || defined(CONFIG_CPU_CORTEX_M0)) && \
-	    _thread_offset_to_basepri > 124
+#if (defined(CONFIG_CPU_CORTEX_M0PLUS) || defined(CONFIG_CPU_CORTEX_M0) || \
+	defined(CONFIG_CPU_CORTEX_M1)) && _thread_offset_to_basepri > 124
```
Comment (ioannisg) [inline arch/arm/core/aarch32/swap_helper.S:174]: In theory, this could be ARMV6_M_ARMV8_M_BASELINE, but I have not tested it in Cortex-M23 :)

File: arch/arm/core/aarch32/cortex_m/Kconfig (hunk 5 above)
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply arch/arm/core/aarch32/cortex_m/Kconfig:30]: I see your point. How do you propose we handle the NVIC, then? It too depends on `OS Extensions`.

**Comment by ioannisg** [comment]
So, the SoC-level .dtsi should include the definitions of the SoC nodes.
In your case, everything in the `soc {` node (gpio, uart, etc.) should go in a file, like: `zephyr/dts/arm/_vendor_/soc.dtsi`

You are right about MuSca and MPS2, but all the rest of cortex-m SoCs have a dts/ .dtsi header.

File: arch/arm/core/aarch32/cortex_m/Kconfig (hunk 5 above)
[the remark this answers is not in the document]
Comment (ioannisg) [reply arch/arm/core/aarch32/cortex_m/Kconfig:30]: There is no Kconfig option related to NVIC, as far as I remember. So from Kconfig's perspective we are fine, for now. 
In Cortex-M, you can use `CONFIG_NUM_IRQS = 0`, to signify a CPU with no NVIC (so the core has only the ARM exceptions).

**Comment by henrikbrixandersen** [comment]
> So, the SoC-level .dtsi should include the definitions of the SoC nodes.
> In your case, everything in the `soc {` node (gpio, uart, etc.) should go in a file, like: `zephyr/dts/arm/_vendor_/soc.dtsi`
> 
> You are right about MuSca and MPS2, but all the rest of cortex-m SoCs have a dts/ .dtsi header.

The "SoC" here only consists of a soft-core CPU. All SoC peripherals (except for the NVIC, if enabled) are provided by the custom, application-specific, programmable logic design. This is for FPGAs, not ASICs.

One such design is the ARM Cortex-M1 DesignStart FPGA for Xilinx reference implementation, but it is just that - a reference implementation for demo purposes, specific to one single board (the Digilent Arty A7). It is not suitable for real applications.

A real world use-case would need a custom board with a custom DTS describing the peripherals implemented in the programmable logic of that specific design.

File: arch/arm/core/aarch32/cortex_m/Kconfig (hunk 5 above)
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply arch/arm/core/aarch32/cortex_m/Kconfig:30]: Compiling with `CONFIG_NUM_IRQS=0` gives compilation errors in `isr_tables.c`. I suggest leaving it at 32 on the SoC level.

[the remark this answers is not in the document]
Comment (ioannisg) [reply arch/arm/core/aarch32/cortex_m/Kconfig:30]: OK, this should be fixed. In my opinion CONFIG_NUM_IRQS=0 should be a valid configuration. Maybe we should open a ticket.

File: tests/arch/arm/arm_interrupt/src/arm_interrupt.c (hunk 2 above)
Comment (henrikbrixandersen) [reply tests/arch/arm/arm_interrupt/src/arm_interrupt.c:98]: @ioannisg I have updated the comments in the test cases with your suggestion. Thanks.

File: arch/arm/core/aarch32/swap_helper.S (hunk 6 above)
Comment (henrikbrixandersen) [reply arch/arm/core/aarch32/swap_helper.S:174]: Yeah, I wondered why it was CPU specific instead of baseline specific. I will leave it as-is until somebody validates it on Cortex-M23.

File: arch/arm/core/aarch32/cortex_m/Kconfig (hunk 5 above)
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply arch/arm/core/aarch32/cortex_m/Kconfig:30]: @ioannisg I have opened #27490 to track that issue.

[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply arch/arm/core/aarch32/cortex_m/Kconfig]: After discussing this on Slack, I have moved `CONFIG_CPU_CORTEX_M_HAS_SYSTICK=y` to the board level and removed `CONFIG_NUM_IRQS=32` default from the SoC level.

File: soc/arm/arm/designstart/Kconfig.defconfig.cortex_m1 [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2020 Henrik Brix Andersen <henrik@brixandersen.dk>
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_ARM_DESIGNSTART_FPGA_CORTEX_M1
+
+config SOC
+	default "cortex_m1"
```
Comment (ioannisg) [inline soc/arm/arm/designstart/Kconfig.defconfig.cortex_m1:7]: This could be designstart_cortex_m1, to make it more specific

**Review by ioannisg — APPROVED** [review]
legitimate now, thanks

**Comment by ioannisg** [comment]
@MaureenHelm @galak @stephanosio one more review here would be needed

File: soc/arm/arm/designstart/Kconfig.defconfig.cortex_m1 (hunk 7 above)
Comment (henrikbrixandersen) [reply soc/arm/arm/designstart/Kconfig.defconfig.cortex_m1]: Fixed.

File: boards/arm/arty/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+  if(NOT ${BIN2HEX} STREQUAL BIN2HEX-NOTFOUND)
+    set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+      COMMAND ${BIN2HEX}
+      ARGS --bit-width 32
+      ${PROJECT_BINARY_DIR}/${CONFIG_KERNEL_BIN_NAME}.bin
+      ${PROJECT_BINARY_DIR}/${CONFIG_KERNEL_BIN_NAME}.mem
+      WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+      )
+  endif()
```
Comment (tejlmand) [inline boards/arm/arty/CMakeLists.txt:22]: just wondering.
a user not having `bin2hex` installed, and have `CONFIG_BUILD_OUTPUT_BIN=y` might wonder why no hex file is being produced.

Could you add a message that `hex-file` output can not be produced. (build should not fail imho).
Either at the CMake configure stage or as an extra post build command.

One thing that favors build time, is that user will be notified on each `ninja/make` invocation that a hex file cannot be produced.
Also if user uses `west build` then a CMake message might not be noticed, whereas a build time message is more likely to get attention.

As example:
```
bin2hex could not be found, hex file cannot be produce.
Please install bin2hex or disable hex-file output with `CONFIG_BUILD_OUTPUT_BIN=n`
```

**Review by tejlmand — CHANGES_REQUESTED** [review]
Looks good, only a minor suggestion to help users not having `bin2hex` installed.

File: boards/arm/arty/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,23 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if((CONFIG_BOARD_ARTY_A7_ARM_DESIGNSTART_M1) AND (CONFIG_BUILD_OUTPUT_BIN))
```
Comment (tejlmand) [inline boards/arm/arty/CMakeLists.txt:3]: should this have been ?:
```suggestion
if((CONFIG_BOARD_ARTY_A7_ARM_DESIGNSTART_M1) AND
   (CONFIG_BUILD_OUTPUT_BIN) AND
   (CONFIG_BUILD_OUTPUT_HEX)
)
```

after all, what is the point in looking for `bin2hex` if `CONFIG_BUILD_OUTPUT_HEX=n`.

**Review by tejlmand — CHANGES_REQUESTED** [review]
should hex file (bin2hex) output be dependent on `CONFIG_BUILD_OUTPUT_HEX` ?

File: boards/arm/arty/CMakeLists.txt (hunk 8 above)
Comment (henrikbrixandersen) [reply boards/arm/arty/CMakeLists.txt]: @tejlmand Thank you for your feedback. I have added a couple of messages to the board `CMakeLists.txt` file. I opted to output these during configuration, since most people will just use the .elf file for development (and the .mem file for deployment).

File: boards/arm/arty/CMakeLists.txt (hunk 9 above)
Comment (henrikbrixandersen) [reply boards/arm/arty/CMakeLists.txt:3]: No, it needs to depend on `CONFIG_BUILD_OUTPUT_BIN` since the .bin file is used as input for generating the .mem file. While the .mem file is a hex dump of the memory, it is in a different format (Verilog memory hex dump) than the hex file (intel hex format) generated when enabling `CONFIG_BUILD_OUTPUT_HEX`.

Comment (tejlmand) [reply boards/arm/arty/CMakeLists.txt:3]: Thanks for explaining.

Of course the bin file is still a dependency, which is why I kept it in my code suggestion ^^^.

I'm also aware that the verilog hex format is different from the intel hex format, but I assume the intel hex format is of no interest on this target, therefore I simply wondered if `CONFIG_BUILD_OUTPUT_HEX` should be used to indicate verilog hex format on this target.

Or maybe we should provide an additional setting here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/Kconfig.zephyr#L295-L319
after all, we do support `intel hex`, `bin`, `s19`, `exe` so maybe a verilog format would make sense for more targets in future.

But I let you decide on this.

The updated messages looks good.

File: boards/arm/arty/CMakeLists.txt (hunk 8 above)
Comment (tejlmand) [reply boards/arm/arty/CMakeLists.txt]: thanks.

**Review by tejlmand — APPROVED** [review]
lgtm

File: boards/arm/arty/CMakeLists.txt (hunk 9 above)
Comment (henrikbrixandersen) [reply boards/arm/arty/CMakeLists.txt:3]: I plan on submitting proper, tree-wide support for Verilog memory hex dump format outputs once the binutils Verilog bfd bug (https://sourceware.org/bugzilla/show_bug.cgi?id=25202) is resolved. I have a branch ready for it here: https://github.com/henrikbrixandersen/zephyr/tree/verilog-memory-hex-dump

The board-specific solution here is just a temporary work-around for now. I would rather not interfere with `CONFIG_BUILD_OUTPUT_HEX` since I believe that might confuse users even more. The requirement on `CONFIG_BUILD_OUTPUT_BIN` for the Verilog .mem file generation has been added to the board documentation.

Comment (tejlmand) [reply boards/arm/arty/CMakeLists.txt:3]: sounds great.

**Review by carlescufi — APPROVED** [review]
This looks great, nice job.

**Comment by carlescufi** [comment]
@MaureenHelm would you like to take a look?

**Review by tyrantlizard77 — APPROVED** [review]
not an expert on this platform but this PR LGTM.

**Review by MaureenHelm — APPROVED** [review]
